compiler/src/language_server/definition.re

```diff
@@ -54,37 +54,64 @@ let send_definition =
   );
 };
 
-let process =
-    (
-      ~id: Protocol.message_id,
-      ~compiled_code: Hashtbl.t(Protocol.uri, Lsp_types.code),
-      ~documents: Hashtbl.t(Protocol.uri, string),
-      params: RequestParams.t,
-    ) => {
-  switch (Hashtbl.find_opt(compiled_code, params.text_document.uri)) {
-  | None => send_no_result(~id)
-  | Some({program, sourcetree}) =>
-    let results = Sourcetree.query(params.position, sourcetree);
+let rec find_definition =
+        (
+          ~depth=0,
+          sourcetree: Sourcetree.sourcetree,
+          position: Protocol.position,
+        ) => {
+  let results = Sourcetree.query(position, sourcetree);
 
+  let result =
     switch (results) {
     | [Value({definition}), ..._]
     | [Pattern({definition}), ..._]
     | [Type({definition}), ..._]
     | [Declaration({definition}), ..._]
     | [Module({definition}), ..._] =>
       switch (definition) {
-      | None => send_no_result(~id)
+      | None => None
       | Some(loc) =>
         let uri = Utils.filename_to_uri(loc.loc_start.pos_fname);
-
-        send_definition(
-          ~id,
-          ~range=Utils.loc_to_range(loc),
-          ~target_uri=uri,
-          Utils.loc_to_range(loc),
-        );
+        Some((loc, uri));
       }
-    | _ => send_no_result(~id)
+    | _ => None
+    };
+  switch (result) {
+  | None =>
+    if (depth == 0 && position.character > 0) {
+      find_definition(
+        ~depth=1,
```

Comment:
You are using `~depth` with an integer that is only ever `1` or `0`. That seems feels like it means it should be `bool` but that is probably not clear enough to explain the purpose so maybe you should make variants for the states this function can handle.

Reply:
I fixed this by using a better variable name and adding a comment. as I didnt feel that a variant provided enough context as to what we were doing.

Reply:
Reading your comment helps, but I think this should be variants that wrap the `Position.position` and probably called `Forwards` and `Backwards` or something. Bool flags are almost always bad design.

Reply:
I switched to a variant, named `Forwards, Backwards`